### ISSUE TITLE
Lint reading an option and then overwriting it with `None`

### ISSUE THREAD

**Issue by oli-obk** [Manishearth/rust-clippy#1853]
```rust
let old = some_option;
... code not referencing `some_option`
some_option = None;
// use `old`
```

should be 

```rust
let old = some_option.take();
```

Some care must be taken not to change behaviour. Accessing `old` and changing control flow (break, return) before `some_option = None` would be a legitimate case that the lint should not trigger on.
